### PR TITLE
[SPARK-21091][SQL] Move constraint code into QueryPlanConstraints

### PR DESCRIPTION
**Pull request by rxin**
## What changes were proposed in this pull request?
This patch moves constraint related code into a separate trait QueryPlanConstraints, so we don't litter QueryPlan with a lot of constraint private functions.

## How was this patch tested?
This is a simple move refactoring and should be covered by existing tests.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlanConstraints.scala [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    case _: NullIntolerant => expr.children.flatMap(scanNullIntolerantAttribute)
+    case _ => Seq.empty[Attribute]
+  }
+
+  // Collect aliases from expressions of the whole tree rooted by the current QueryPlan node, so
+  // we may avoid producing recursive constraints.
+  private lazy val aliasMap: AttributeMap[Expression] = AttributeMap(
+    expressions.collect {
+      case a: Alias => (a.toAttribute, a.child)
+    } ++ children.flatMap(_.asInstanceOf[QueryPlanConstraints[PlanType]].aliasMap))
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlanConstraints.scala]: this is the only weird part; the rest are straightforward copy.

I couldn't figure out how to get the compiler to past without the explicit cast.

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlanConstraints.scala]: what if we do `trait QueryPlanConstraints[PlanType <: QueryPlan[PlanType]] extends QueryPlan[PlanType]`?

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlanConstraints.scala]: that would create a cyclic hierarchy ...

**Comment by gatorsmile** [comment]
retest this please

**Comment by gatorsmile** [comment]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlanConstraints.scala [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+
+  /**
+   * An [[ExpressionSet]] that contains invariants about the rows output by this operator. For
+   * example, if this set contains the expression `a = 2` then that expression is guaranteed to
+   * evaluate to `true` for all rows produced.
+   */
+  lazy val constraints: ExpressionSet =
+    ExpressionSet(
+      validConstraints
+        .union(inferAdditionalConstraints(constraints))
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlanConstraints.scala]: `validConstraints`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlanConstraints.scala [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        .union(constructIsNotNullConstraints(constraints))
```
Comment (gatorsmile) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlanConstraints.scala]: `validConstraints`

**Comment by sameeragarwal** [comment]
Thanks for the cleanup! LGTM modulo Xiao's comment.

**Comment by gatorsmile** [comment]
LGTM

**Comment by rxin** [comment]
Merging in master.

**Pull request closed without merging**